File: cli/agent_system/commands/workflow.py
"""workflow commands: list, id, run, logs."""

from __future__ import annotations

import typer
from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from agent_system import api

app = typer.Typer(help="Workflow commands.")
console = Console()


def _ts(val) -> str:
    if not val:
        return ""
    try:
        from datetime import datetime, timezone
        if isinstance(val, (int, float)):
            dt = datetime.fromtimestamp(val / 1000, tz=timezone.utc)
        else:
            dt = datetime.fromisoformat(str(val).replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d %H:%M")
    except Exception:
        return str(val)


@app.command("list")
def list_workflows():
    """List your workflows."""
    workflows = api.get("/api/v1/workflow")
    if not workflows:
        console.print("[dim]No workflows found.[/dim]")
        return

    table = Table(title="Workflows", show_lines=True)
    table.add_column("ID", style="cyan", no_wrap=True)
    table.add_column("Name", style="bold white")
    table.add_column("Pattern", style="yellow")
    table.add_column("Agents", justify="right")
    table.add_column("Created", style="dim")

    for w in workflows:
        agents = w.get("agents") or []
        table.add_row(
            w.get("id", ""),
            w.get("name", ""),
            w.get("agentPattern", ""),
            str(len(agents)),
            _ts(w.get("createdAt")),
        )

    console.print(table)


@app.command("id")
def get_workflow(workflow_id: str = typer.Argument(..., help="Workflow ID")):
    """Show details of a workflow including its agents."""
    w = api.get(f"/api/v1/workflow/{workflow_id}")

    console.print(Panel(
        f"[bold]{w.get('name', '')}[/bold]\n"
        f"[dim]{w.get('description', '')}[/dim]\n\n"
        f"Pattern: [yellow]{w.get('agentPattern', '')}[/yellow]   "
        f"Exec mode: [cyan]{w.get('teamExecMode', 'N/A')}[/cyan]",
        title=f"Workflow [cyan]{workflow_id}[/cyan]",
        border_style="cyan",
    ))

    agents = api.get(f"/api/v1/workflow/{workflow_id}/agents")
    if not agents:
        console.print("[dim]No agents configured.[/dim]")
        return

    table = Table(title="Agents", show_lines=True)
    table.add_column("#", style="dim", width=4)
    table.add_column("Name", style="bold white")
    table.add_column("Role", style="yellow")
    table.add_column("Tools", style="cyan")
    table.add_column("System prompt", style="dim")

    for a in sorted(agents, key=lambda x: x.get("orderIndex", 0)):
        tools = ", ".join(a.get("tools") or []) or "—"
        prompt = (a.get("systemPrompt") or "")[:50]
        if len(a.get("systemPrompt") or "") > 50:
            prompt += "…"
        table.add_row(
            str(a.get("orderIndex", "")),
            a.get("name", ""),
            a.get("role", ""),
            tools,
            prompt,
        )

    console.print(table)


@app.command("run")
def run_workflow(
    workflow_id: str = typer.Argument(..., help="Workflow ID"),
    input: str = typer.Argument(..., help="User input / task description"),
    notify: bool = typer.Option(False, "--notify", "-n", help="Email notification on completion"),
):
    """Trigger a workflow run."""
    with console.status("[cyan]Starting run…[/cyan]"):
        result = api.post(
            f"/api/v1/workflow/{workflow_id}/runs",
            {"userInput": input, "emailNotify": notify},
        )
    run_id = result.get("runId", "")
    console.print(Panel(
        f"Run started: [bold cyan]{run_id}[/bold cyan]\n\n"
        f"Check logs with:\n  [bold]agent-system workflow logs {run_id}[/bold]",
        title="[green]Run started[/green]",
        border_style="green",
    ))


@app.command("runs")
def list_runs(workflow_id: str = typer.Argument(..., help="Workflow ID")):
    """List recent runs for a workflow."""
    runs = api.get(f"/api/v1/workflow/{workflow_id}/runs")
    if not runs:
        console.print("[dim]No runs found.[/dim]")
        return

    table = Table(title=f"Runs for workflow {workflow_id}", show_lines=True)
    table.add_column("Run ID", style="cyan", no_wrap=True)
    table.add_column("Status", style="bold")
    table.add_column("Started", style="dim")
    table.add_column("Finished", style="dim")

    status_colors = {"COMPLETED": "green", "FAILED": "red", "RUNNING": "yellow", "PENDING": "dim"}
    for r in runs:
        status = r.get("status", "UNKNOWN")
        color  = status_colors.get(status, "white")
        table.add_row(
            r.get("id", ""),
            Text(status, style=color),
            _ts(r.get("startedAt")),
            _ts(r.get("finishedAt")),
        )

    console.print(table)


@app.command("logs")
def get_logs(run_id: str = typer.Argument(..., help="Run ID")):
    """Show logs for a completed workflow run."""
    logs = api.get(f"/api/v1/workflow/runs/{run_id}/logs")
    if not logs:
        console.print("[dim]No logs found for this run.[/dim]")
        return

    level_styles = {"ERROR": "bold red", "WARN": "yellow", "INFO": "cyan", "DEBUG": "dim"}
    for entry in logs:
        level   = entry.get("level", "INFO")
        agent   = entry.get("agentName", "")
        message = entry.get("message", "")
        ts      = _ts(entry.get("createdAt"))
        style   = level_styles.get(level, "white")
        console.print(f"[dim]{ts}[/dim]  [{style}]{level:<5}[/{style}]  [bold]{agent}[/bold]  {message}")

File: cli/cmd/workflow.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"text/tabwriter"

	"cli/api"

	"github.com/spf13/cobra"
)

type workflow struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	AgentPattern string `json:"agentPattern"`
	TeamExecMode string `json:"teamExecMode"`
	Description  string `json:"description"`
	Agents       []any  `json:"agents"`
	CreatedAt    any    `json:"createdAt"`
}

type wfAgent struct {
	OrderIndex   int      `json:"orderIndex"`
	Name         string   `json:"name"`
	Role         string   `json:"role"`
	Tools        []string `json:"tools"`
	SystemPrompt string   `json:"systemPrompt"`
}

type wfRun struct {
	ID         string `json:"id"`
	Status     string `json:"status"`
	StartedAt  any    `json:"startedAt"`
	FinishedAt any    `json:"finishedAt"`
}

type wfLog struct {
	Level     string `json:"level"`
	AgentName string `json:"agentName"`
	Message   string `json:"message"`
	CreatedAt any    `json:"createdAt"`
}

var workflowCmd = &cobra.Command{
	Use:   "workflow",
	Short: "Workflow commands",
}

func init() {
	workflowCmd.AddCommand(&cobra.Command{
		Use:   "list",
		Short: "List your workflows",
		RunE: func(_ *cobra.Command, _ []string) error {
			raw, err := api.Get("/api/v1/workflow")
			if err != nil {
				return err
			}
			var wfs []workflow
			if err := json.Unmarshal(raw, &wfs); err != nil {
				return err
			}
			if len(wfs) == 0 {
				fmt.Println(d("No workflows found."))
				return nil
			}
			tw := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
			fmt.Fprintln(tw, "ID\tNAME\tPATTERN\tAGENTS\tCREATED")
			for _, wf := range wfs {
				fmt.Fprintf(tw, "%s\t%s\t%s\t%d\t%s\n",
					wf.ID, short(wf.Name, 30), wf.AgentPattern, len(wf.Agents), fmtTime(wf.CreatedAt))
			}
			tw.Flush()
			return nil
		},
	})

	workflowCmd.AddCommand(&cobra.Command{
		Use:   "id <workflow-id>",
		Short: "Show workflow details including agents",
		Args:  cobra.ExactArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			raw, err := api.Get("/api/v1/workflow/" + args[0])
			if err != nil {
				return err
			}
			var wf workflow
			if err := json.Unmarshal(raw, &wf); err != nil {
				return err
			}
			fmt.Printf("\n%s\n%s\n", b(wf.Name), d(wf.Description))
			fmt.Printf("Pattern: %s   Mode: %s\n\n", y(wf.AgentPattern), c(wf.TeamExecMode))

			agRaw, err := api.Get("/api/v1/workflow/" + args[0] + "/agents")
			if err != nil {
				return err
			}
			var agents []wfAgent
			if err := json.Unmarshal(agRaw, &agents); err != nil {
				return err
			}
			if len(agents) == 0 {
				fmt.Println(d("No agents configured."))
				return nil
			}
			tw := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
			fmt.Fprintln(tw, "#\tNAME\tROLE\tTOOLS\tSYSTEM PROMPT")
			for _, a := range agents {
				tools := "—"
				if len(a.Tools) > 0 {
					tools = strings.Join(a.Tools, ",")
				}
				fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\n",
					a.OrderIndex, a.Name, a.Role, tools, short(a.SystemPrompt, 50))
			}
			tw.Flush()
			return nil
		},
	})

	runCmd := &cobra.Command{
		Use:   "run <workflow-id> <input>",
		Short: "Trigger a workflow run",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			notify, _ := cmd.Flags().GetBool("notify")
			fmt.Fprintln(os.Stderr, d("Starting run…"))
			raw, err := api.Post("/api/v1/workflow/"+args[0]+"/runs", map[string]any{
				"userInput":   args[1],
				"emailNotify": notify,
			})
			if err != nil {
				return err
			}
			var res struct {
				RunID string `json:"runId"`
			}
			if err := json.Unmarshal(raw, &res); err != nil {
				return err
			}
			fmt.Printf("%s Run started: %s\n", g("✓"), b(res.RunID))
			fmt.Printf("Check logs: %s\n", c("agent-system workflow logs "+res.RunID))
			return nil
		},
	}
	runCmd.Flags().BoolP("notify", "n", false, "Email notification on completion")
	workflowCmd.AddCommand(runCmd)

	workflowCmd.AddCommand(&cobra.Command{
		Use:   "runs <workflow-id>",
		Short: "List recent runs for a workflow",
		Args:  cobra.ExactArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			raw, err := api.Get("/api/v1/workflow/" + args[0] + "/runs")
			if err != nil {
				return err
			}
			var runs []wfRun
			if err := json.Unmarshal(raw, &runs); err != nil {
				return err
			}
			if len(runs) == 0 {
				fmt.Println(d("No runs found."))
				return nil
			}
			tw := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
			fmt.Fprintln(tw, "RUN ID\tSTATUS\tSTARTED\tFINISHED")
			for _, ru := range runs {
				fmt.Fprintf(tw, "%s\t%s\t%s\t%s\n",
					ru.ID, ru.Status, fmtTime(ru.StartedAt), fmtTime(ru.FinishedAt))
			}
			tw.Flush()
			return nil
		},
	})

	workflowCmd.AddCommand(&cobra.Command{
		Use:   "logs <run-id>",
		Short: "Show logs for a completed workflow run",
		Args:  cobra.ExactArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			raw, err := api.Get("/api/v1/workflow/runs/" + args[0] + "/logs")
			if err != nil {
				return err
			}
			var logs []wfLog
			if err := json.Unmarshal(raw, &logs); err != nil {
				return err
			}
			if len(logs) == 0 {
				fmt.Println(d("No logs found."))
				return nil
			}
			for _, entry := range logs {
				lvl := fmt.Sprintf("%-5s", entry.Level)
				var lvlColor string
				switch entry.Level {
				case "ERROR":
					lvlColor = r(lvl)
				case "WARN":
					lvlColor = y(lvl)
				case "INFO":
					lvlColor = c(lvl)
				default:
					lvlColor = d(lvl)
				}
				fmt.Printf("%s  %s  %s  %s\n",
					d(fmtTime(entry.CreatedAt)), lvlColor, b(entry.AgentName), entry.Message)
			}
			return nil
		},
	})

	rootCmd.AddCommand(workflowCmd)
}
